segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/LogStorage.java

```diff
@@ -0,0 +1,201 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.impl.bookkeeperstorage;
+
+import io.pravega.common.util.ImmutableDate;
+import io.pravega.segmentstore.contracts.BadOffsetException;
+import java.nio.ByteBuffer;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.Getter;
+import org.apache.curator.framework.api.transaction.CuratorOp;
+
+/*
+ * Storage Log represents a single segment. A segment is rolled over whenever a ownership change is observed.
+ *
+ **/
+class LogStorage {
+
+    private final LogStorageManager manager;
```

Comment:
This class should not have a pointer to `LogStorageManager`. It really should be a pure data class, managed by `LogStorageManager`. 

In addition, I would make this class mostly immutable (check out `LogMetadata` in the BK Tier1). Any changes to it would result in a new object, which is a copy of the old one. That way we resolve most of the concurrency issues we have to deal with here.

Reply:
Fixed. Made `LogStorage` a data class.